File: libgomp/testsuite/libgomp.fortran/omp-get-device-distances.f90
use omp_lib
implicit none (type, external)
integer :: devs(1), dists(1)
integer, allocatable :: a_devs(:), a_dists(:)
integer :: i

! GCC specific: invalid numbers return -1
devs(1) = omp_invalid_device; dists(1) = -99
call omp_get_device_distances (size(devs), devs, dists)
if (devs(1) /= omp_invalid_device .or. dists(1) /= -1) stop 1

devs(1) = -123; dists(1) = -99
call omp_get_device_distances (size(devs), devs, dists)
if (devs(1) /= -123 .or. dists(1) /= -1) stop 2

devs(1) = omp_get_num_devices () + 1; dists(1) = -99
call omp_get_device_distances (size(devs), devs, dists)
if (devs(1) /= omp_get_num_devices () + 1 .or. dists(1) /= -1) stop 3

do i = omp_initial_device, omp_get_num_devices ()
  devs(1) = i; dists(1) = -99
  call omp_get_device_distances (size(devs), devs, dists)
  if (devs(1) /= i) stop 4
  ! Host == 0 per definition
  ! GCC specific: All others (unknown or not) should have > 0
  ! the spec only requires >= 0
  if (i == omp_initial_device .or. i == omp_get_num_devices ()) then
    if (dists(1) /= 0) stop 5
  else
    if (dists(1) <= 0) stop 6
  end if
end do

allocate(a_devs(omp_initial_device:omp_get_num_devices ()))
a_devs = [(i, i = omp_initial_device, omp_get_num_devices ())]
allocate(a_dists, mold=a_devs)
a_dists = -99

call omp_get_device_distances (size(a_devs), a_devs, a_dists)

do i = lbound (a_devs, 1), ubound (a_devs, 1)
  if (a_devs(i) /= i) stop 7
  ! Host == 0 per definition
  ! GCC specific: All others (unknown or not) should have > 0
  if (i == omp_initial_device .or. i == omp_get_num_devices ()) then
    if (a_dists(i) /= 0) stop 8
  else
    if (a_dists(i) <= 0) stop 9
  end if
  print '(a,i3,a,i0)', 'device', i, ': distance ', a_dists(i)
end do

end

File: libgomp/testsuite/libgomp.c/omp-get-device-distances.c
#include <omp.h>

int
main ()
{
  int dev, dist;
  int dev2, dist2;

  // GCC specific: invalid numbers return -1
  dev = omp_invalid_device; dist = -99;
  omp_get_device_distances (1, &dev, &dist);
  if (dev != omp_invalid_device || dist != -1)
    __builtin_abort ();

  dev = -99; dist = -99;
  omp_get_device_distances (1, &dev, &dist);
  if (dev != -99 || dist != -1)
    __builtin_abort ();

  dev = omp_get_num_devices () + 1; dist = -99;
  omp_get_device_distances (1, &dev, &dist);
  if (dev != omp_get_num_devices () + 1 || dist != -1)
    __builtin_abort ();

  // The following two need to yield idential results
  dev = omp_default_device; dist = -99;
  omp_get_device_distances (1, &dev, &dist);
  if (dev != omp_default_device || dist < 0)
    __builtin_abort ();

  dev2 = omp_get_default_device (); dist2 = -99;
  omp_get_device_distances (1, &dev2, &dist2);
  if (dev2 != omp_get_default_device () || dist2 != dist)
    __builtin_abort ();

  // The following two need to yield idential results
  dev = omp_initial_device; dist = -99;
  omp_get_device_distances (1, &dev, &dist);
  if (dev != omp_initial_device || dist != 0)
    __builtin_abort ();

  dev2 = omp_get_num_devices (); dist2 = -99;
  omp_get_device_distances (1, &dev2, &dist2);
  if (dev2 != omp_get_num_devices () || dist2 != 0)
    __builtin_abort ();

  for (int i = omp_initial_device; i <= omp_get_num_devices (); i++)
    {
      dev = i; dist = -99;
      omp_get_device_distances (1, &dev, &dist);
      if (dev != i)
	__builtin_abort ();
      // Host == 0 per definition
      // GCC specific: All others (unknown or not) should have > 0
      // the spec only requires >= 0
      if (i == omp_initial_device || i == omp_get_num_devices ())
	{
	  if (dist != 0)
	    __builtin_abort ();
	}
      else
	{
	  if (dist <= 0)
	    __builtin_abort ();
	}
     }

  int *devs, *dists;
  int size = 1 + 1 + omp_get_num_devices ();
  devs = (int *) __builtin_malloc (sizeof (int) * size * size);
  dists = (int *) __builtin_malloc (sizeof (int) * size * size);
  for (int i = 0; i < size; i++)
    {
      int dev = i - 1; // -1 (initial), 0, ... num_dev
      devs[i] = dev;
      dists[i] = -99;
    }

  omp_get_device_distances (size, devs, dists);

  for (int i = 0; i < size; i++)
    {
      int dev = i - 1; // -1 (initial), 0, ... num_dev
      if (devs[i] != dev)
	__builtin_abort ();
      // Host == 0 per definition
      // GCC specific: All others (unknown or not) should have > 0
      // the spec only requires >= 0
      if (devs[i] == omp_initial_device || devs[i] == omp_get_num_devices ())
	{
	  if (dists[i] != 0)
	    __builtin_abort ();
	}
      else
	{
	  if (dists[i] <= 0)
	    __builtin_abort ();
	}
      __builtin_printf ("Device%3d: distance %d\n", dev, dists[i]);
    }
}
